File: solutions/python/code-carl/ch10_monotone_stack/t1_lc0739_daily_temperatures.py
"""LeetCode 739. Daily Temperatures
https://leetcode.cn/problems/daily-temperatures/
"""

from typing import List


class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        stack: List[int] = []
        ans: List[int] = [0] * len(temperatures)

        for idx, num in enumerate(temperatures):
            while stack and num > temperatures[stack[-1]]:
                popIdx = stack.pop()
                ans[popIdx] = idx - popIdx
            stack.append(idx)

        return ans


if __name__ == "__main__":
    solution = Solution()
    assert solution.dailyTemperatures([73, 74, 75, 71, 69, 72, 76, 73]) == [1, 1, 4, 2, 1, 1, 0, 0]

    assert solution.dailyTemperatures([73, 74, 75, 71, 69, 72, 76, 73]) == [1, 1, 4, 2, 1, 1, 0, 0]

    print("All tests passed.")

File: solutions/java/code-carl/ch10_monotone_stack/T1_LC0739_DailyTemperatures.java
/**
 * Practice template for:
 * - LeetCode 739. Daily Temperatures
 *   https://leetcode.cn/problems/daily-temperatures/
 */
import java.util.ArrayDeque;
import java.util.Deque;

/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/28 16:33
 * @comment
 */
public class T1_LC0739_DailyTemperatures {
    public int[] dailyTemperatures(int[] temperatures) {
        Deque<Integer> stack = new ArrayDeque<>();
        int[] res = new int[temperatures.length];
        for (int i  = 0; i < temperatures.length; i++) {
            while (!stack.isEmpty() && temperatures[i] > temperatures[stack.peek()]) {
                Integer idx = stack.pop();
                res[idx] = i - idx;
            }
            stack.push(i);
        }
        return res;
    }
}
